### ISSUE TITLE
Add an API for custom test method parameter injection

### ISSUE THREAD

**Issue by Crazyjavahacking** [testng-team/testng#612]
I need a way to inject a custom parameter into TestNG's test methods. **The parameter is my own object constructed dynamically at runtime from test method annotations**. The custom parameter is not known to TestNG, it's my own type, e.g.:

```
@com.mypackage.MyAnnotation("someValue)"
@org.testng.Test
public void test(MyCustomObject val) {
    ...
}
```

It looks like right now I cannot do it easily.

Can we add such functionality/SPI? What is the current easiest way to achieve that?

**Comment by Crazyjavahacking** [comment]
**I don't want to use @DataProvider annotation.**

**Comment by cbeust** [comment]
Why don't you want to use @DataProvider, it's the easiest approach by far
(just a few lines of code).

-- 
Cédric

**Comment by Crazyjavahacking** [comment]
The problem is that the object I want to inject is dynamically created based on the data inside annotations on test method.

In my initial example, I need to parse the content of `@MyAnnotation` and based on that I need to create an instance of `MyCustomObject`.

**Comment by cbeust** [comment]
What is stopping you from doing all that in the data provider?


-- 
Cédric

**Comment by Crazyjavahacking** [comment]
The functionality is not really parameterized testing and I also don't want to use data providers, e.g.:
```
@Columns("col1 | col2 || val)"
@Rows  ({"a    | b    || c",
         "d    | e    || f})
@org.testng.Test
public void test1(MyCustomObject val) {
    // MyCustomerObject is created from the annotation values
    // some menipulation with val ...
}

@Columns("col1 | col2 || val)"
@Rows  ({"1    | 2    || a",
         "3    | 3    || b})
@org.testng.Test
public void test1(MyCustomObject val) {
    // MyCustomerObject is created from the annotation values
    // some menipulation with val ...
}
```
so I don't think data providers are a good approach here

**Comment by g13013** [comment]
Would be very useful I have the same needs

**Comment by juherr** [comment]
The idea is interesting but as @cbeust said a long time ago, a data provider is a good workaround.
According to the previous sample:

```java
@Columns("col1 | col2 || val)")
@Rows  ({    "a    | b      || c",
                    "d    | e      || f"})
@Test(dataProvider = "dp")
public void test1(MyCustomObject val) {
    // MyCustomerObject is created from the annotation values
    // some menipulation with val ...
}

@Columns("col1 | col2 || val)"
@Rows(      {"1    | 2      || a",
                    "3    | 3      || b"})
@Test(dataProvider = "dp")
public void test1(MyCustomObject val) {
    // MyCustomerObject is created from the annotation values
    // some menipulation with val ...
}

@DataProvider
public static MyCustomObject[] dp(Method m) {
    Columns colums = m.getAnnotation(Columns.class);
    Rows rows = m.getAnnotation(Rows.class);
    MyCustomObject[] results = new MyCustomObject[rows.values().length];
    for (int i=0; i < rows.values().length; i++) {
        String value = rows.values()[i];
        // extract params with custom business
        results[i] = new MyCustomObject(/* params */);
    }
    return results;
}
```

**Comment by g13013** [comment]
this is true @juherr, i ended doing it this way. But, the only thing that misses is to set a default dataProvider for all tests, in my case, my data provider is generics and passes only if there is certain annotations

i think that  a combination of the dataProvider and adding the ability to set a default provider would do the trick and will open a hole new possibilities

**Comment by juherr** [comment]
@g13013 You mean something like

```java
public class UseMyDataProvider implements IAnnotationTransformer {
  public void transform(ITest annotation, Class testClass, Constructor testConstructor, Method testMethod) {
    if (/* business logic (for example: testMethod has expected annotations) */) {
      annotation.setDataProviderClass(MyExternalClassWithDataProvider.class);
      annotation.setDataProvider("dataprovider name");
    }
  }
}
```

**Comment by gjuillot** [comment]
I would also love to be able to use Guice to inject parameters in tests.
DataProvider + IAnnotationTransformer is an interesting workaround but it lack flexibility (for example it is impossible to set the IAnnotationTransformer via @Listeners)

**Comment by juherr** [comment]
> it is impossible to set the IAnnotationTransformer via `@Listeners`

@gjuillot It could be a feature request. Could you open an issue with that?

**Comment by gjuillot** [comment]
I'll do. But still, the best ever for me would be to have the possibility to use Guice to inject parameters in tests.

**Comment by juherr** [comment]
Verdict: still relevant.

Current TestNG still has only a closed set of native method-parameter injections (`Method`, `ITestContext`, `ITestResult`, `XmlTest`) plus XML parameters/data-provider values; there is no generic resolver SPI on `master` for an arbitrary third-party parameter built from its annotations.

There is now a direct open implementation: PR #3459 (`feat(injection): resolve test method parameters through IParameterResolver`). It introduces a public `IParameterResolver`, registers it through the normal listener mechanisms, composes it with data providers/native injection/varargs/retries, and includes dedicated resolver plus Mockito integration tests. The PR is currently open, non-draft and reported mergeable; it explicitly notes that #612 is substantially addressed. I have not executed its branch or tests in this triage pass, so I am not treating the issue as resolved until the PR is merged.

The historical DataProvider + `IAnnotationTransformer` approach remains a workaround, not the requested SPI. Next step: review and merge #3459 (or revise its public contract); no separate production implementation should be started while that PR remains the active implementation.
